### ISSUE TITLE
Using `pendulum.DateTime` snapshots with network containing links

### ISSUE THREAD

**Issue by toby-coleman** [PyPSA/PyPSA#533]
## Checklist

- [x] I am using the latest [release](https://github.com/pypsa/pypsa/releases). Please indicate: `0.21.3`.

## Describe the Bug

Very similar to #470 but occurs when optimising a network that contains a link.

```python
import pypsa
import pandas as pd
import pendulum

dts = list((pendulum.datetime(2022, 1, 1) - pendulum.datetime(2022, 1, 2)).range("minutes", 30))
# Removing tzinfo is a workaround, as before with #470
# dts = [d.replace(tzinfo=None) for d in (pendulum.datetime(2022, 10, 2) - pendulum.datetime(2022, 10, 1)).range("minutes", 30)]


demand = pd.DataFrame({"demand": [10] * len(dts)}, index=dts)

network = pypsa.Network()
network.set_snapshots(dts)

network.add("Bus", "b1")
network.add("Bus", "b2")

network.add("Link", "l1", bus0="b1", bus1="b2", p_nom=1000, p_max_pu=1, p_min_pu=-1)

network.add("Load", "d1", bus="b1", p_set=demand["demand"])
network.add("Generator", "g1", bus="b2", p_nom=10, p_max_pu=1, p_min_pu=0, marginal_cost=50)

network.optimize(solver_name="highs")
```

## Error Message

*If applicable, paste any terminal output to help illustrating your problem.*
*In some cases it may also be useful to share your list of installed packages: `conda list`.*

```
KeyError                                  Traceback (most recent call last)
Cell In [16], line 1
----> 1 network.optimize(solver_name="highs")

File ~/miniconda3/envs/stack-model/lib/python3.9/site-packages/pypsa/optimization/optimize.py:478, in OptimizationAccessor.__call__(self, *args, **kwargs)
    477 def __call__(self, *args, **kwargs):
--> 478     return optimize(self._parent, *args, **kwargs)

File ~/miniconda3/envs/stack-model/lib/python3.9/site-packages/pypsa/optimization/optimize.py:454, in optimize(n, snapshots, multi_investment_periods, linearized_unit_commitment, model_kwargs, extra_functionality, **kwargs)
    451 status, condition = m.solve(**kwargs)
    453 if status == "ok":
--> 454     assign_solution(n)
    455     assign_duals(n)
    456     post_processing(n)

File ~/miniconda3/envs/stack-model/lib/python3.9/site-packages/pypsa/optimization/optimize.py:262, in assign_solution(n)
    260 for i in ["1"] + additional_linkports(n):
    261     i_eff = "" if i == "1" else i
--> 262     eff = get_as_dense(n, "Link", f"efficiency{i_eff}", sns)
    263     set_from_frame(n, c, f"p{i}", -df * eff)
    264     n.pnl(c)[f"p{i}"].loc[
    265         sns, n.links.index[n.links[f"bus{i}"] == ""]
    266     ] = n.component_attrs["Link"].loc[f"p{i}", "default"]

File ~/miniconda3/envs/stack-model/lib/python3.9/site-packages/pypsa/descriptors.py:180, in get_switchable_as_dense(network, component, attr, snapshots, inds)
    178 vals = np.repeat([df.loc[fixed_i, attr].values], len(snapshots), axis=0)
    179 static = pd.DataFrame(vals, index=snapshots, columns=fixed_i)
--> 180 varying = pnl[attr].loc[snapshots, varying_i]
    182 res = pd.concat([static, varying], axis=1, sort=False).reindex(columns=index)
    183 res.index.name = "snapshot"  # reindex with multiindex does not preserve name

File ~/miniconda3/envs/stack-model/lib/python3.9/site-packages/pandas/core/indexing.py:1067, in _LocationIndexer.__getitem__(self, key)
   1065     if self._is_scalar_access(key):
   1066         return self.obj._get_value(*key, takeable=self._takeable)
-> 1067     return self._getitem_tuple(key)
   1068 else:
   1069     # we by definition only have the 0th axis
   1070     axis = self.axis or 0

File ~/miniconda3/envs/stack-model/lib/python3.9/site-packages/pandas/core/indexing.py:1254, in _LocIndexer._getitem_tuple(self, tup)
   1252 # ugly hack for GH #836
   1253 if self._multi_take_opportunity(tup):
-> 1254     return self._multi_take(tup)
   1256 return self._getitem_tuple_same_dim(tup)

File ~/miniconda3/envs/stack-model/lib/python3.9/site-packages/pandas/core/indexing.py:1205, in _LocIndexer._multi_take(self, tup)
   1189 """
   1190 Create the indexers for the passed tuple of keys, and
   1191 executes the take operation. This allows the take operation to be
   (...)
   1202 values: same type as the object being indexed
   1203 """
   1204 # GH 836
-> 1205 d = {
   1206     axis: self._get_listlike_indexer(key, axis)
   1207     for (key, axis) in zip(tup, self.obj._AXIS_ORDERS)
   1208 }
   1209 return self.obj._reindex_with_indexers(d, copy=True, allow_dups=True)

File ~/miniconda3/envs/stack-model/lib/python3.9/site-packages/pandas/core/indexing.py:1206, in <dictcomp>(.0)
   1189 """
   1190 Create the indexers for the passed tuple of keys, and
   1191 executes the take operation. This allows the take operation to be
   (...)
   1202 values: same type as the object being indexed
   1203 """
   1204 # GH 836
   1205 d = {
-> 1206     axis: self._get_listlike_indexer(key, axis)
   1207     for (key, axis) in zip(tup, self.obj._AXIS_ORDERS)
   1208 }
   1209 return self.obj._reindex_with_indexers(d, copy=True, allow_dups=True)

File ~/miniconda3/envs/stack-model/lib/python3.9/site-packages/pandas/core/indexing.py:1432, in _LocIndexer._get_listlike_indexer(self, key, axis)
   1429 ax = self.obj._get_axis(axis)
   1430 axis_name = self.obj._get_axis_name(axis)
-> 1432 keyarr, indexer = ax._get_indexer_strict(key, axis_name)
   1434 return keyarr, indexer

File ~/miniconda3/envs/stack-model/lib/python3.9/site-packages/pandas/core/indexes/base.py:6111, in Index._get_indexer_strict(self, key, axis_name)
   6108 else:
   6109     keyarr, indexer, new_indexer = self._reindex_non_unique(keyarr)
-> 6111 self._raise_if_missing(keyarr, indexer, axis_name)
   6113 keyarr = self.take(indexer)
   6114 if isinstance(key, Index):
   6115     # GH 42790 - Preserve name from an Index

File ~/miniconda3/envs/stack-model/lib/python3.9/site-packages/pandas/core/indexes/base.py:6171, in Index._raise_if_missing(self, key, indexer, axis_name)
   6169     if use_interval_msg:
   6170         key = list(key)
-> 6171     raise KeyError(f"None of [{key}] are in the [{axis_name}]")
   6173 not_found = list(ensure_index(key)[missing_mask.nonzero()[0]].unique())
   6174 raise KeyError(f"{not_found} not in index")

KeyError: "None of [Int64Index([1641081600000000000, 1641079800000000000, 1641078000000000000,\n            1641076200000000000, 1641074400000000000, 1641072600000000000,\n            1641070800000000000, 1641069000000000000, 1641067200000000000,\n            1641065400000000000, 1641063600000000000, 1641061800000000000,\n            1641060000000000000, 1641058200000000000, 1641056400000000000,\n            1641054600000000000, 1641052800000000000, 1641051000000000000,\n            1641049200000000000, 1641047400000000000, 1641045600000000000,\n            1641043800000000000, 1641042000000000000, 1641040200000000000,\n            1641038400000000000, 1641036600000000000, 1641034800000000000,\n            1641033000000000000, 1641031200000000000, 1641029400000000000,\n            1641027600000000000, 1641025800000000000, 1641024000000000000,\n            1641022200000000000, 1641020400000000000, 1641018600000000000,\n            1641016800000000000, 1641015000000000000, 1641013200000000000,\n            1641011400000000000, 1641009600000000000, 1641007800000000000,\n            1641006000000000000, 1641004200000000000, 1641002400000000000,\n            1641000600000000000, 1640998800000000000, 1640997000000000000,\n            1640995200000000000],\n           dtype='int64', name='snapshot')] are in the [index]"```

**Comment by FabianHofmann** [comment]
Thanks @toby-coleman for the reproducible example this is very helpful. Suggested a fix in #536

**Comment by toby-coleman** [comment]
Thanks @FabianHofmann - last week I opened a PR #534 as well: feel free to merge either of them and close the other!
